### ISSUE TITLE
[DOC REQUEST] Examples to override built-in WPBrowser (and friends) methods

### ISSUE THREAD

**Issue by charlesdeb** [lucatume/wp-browser#507]
**Did you search the documentation first?**

Yep - and the codeception stuff too

**What do you think is missing from the documentation?**

I'd like an example showing me how I can override one of the WPBrowser methods. In my use case, I can't just use `loginAs` because I have a third-party authorisation plugin in place which means the normal WP login stuff is diverted. I have been tearing my hair out trying to figure out how to override your loginAs but I just can't figure out how to do it with codeception. In my case, your current login methods would be fine - I just need to override the `getLoginUrl()` and `loginUrl` methods and variables that your methods are using, but I can't even figure out how to stop codeception from using your ones.

**Do you have any examples of what you'd like to see added to the documentation?**

I'm thinking it would be a tutorial maybe, that shows what might go in the `Acceptance\Helper` or `AcceptanceTester` classes in order to either override your `loginAs` or `getLoginUrl` methods. I need this for both the WPBrowser as well as WPWebDriver modules - although once I've cracked it for one, I'm sure the same technique would apply to the other.

The main problem is that I struggle to understand the Codeception documentation properly - but I assume you must know Codeception very well in order to have written WPBrowser to work on top of it so what I am asking for is likely a 10 minute job for you, whereas I have spent several hours trying to get my head around it. I bet others would find this helpful too.

Many thanks!

**Comment by charlesdeb** [comment]
So, after quite a bit of mucking around, I have figured out a way that seems to work for me, and seems to fit with both the codeception and wpbrowser way of doing things. It requires a few moving parts:

Because the login procedure is different for wpwebdriver and wpbrowser, (and I want some tests with javascript and some without - for speed) I have created `acceptance_wpbrowser.suite.yml` and `acceptance_wpwebdriver.suite.yml` config files. They are pretty similar - but here goes the guts of the ``acceptance_wpwebdriver.suite.yml``

```yml
actor: AcceptanceTester
modules:
  enabled:
    - WPDb
    - \Helper\Acceptance
    - \Helper\AcceptanceWPWebDriver:
      depends: WPWebDriver
  config:
    WPDb:
      dsn: '%TEST_SITE_DB_DSN%'
      user: '%TEST_SITE_DB_USER%'
      password: '%TEST_SITE_DB_PASSWORD%'
      dump: 'tests/_data/dump.sql'
      #import the dump before the tests; this means the test site database will be repopulated before the tests.
      populate: true
      # re-import the dump between tests; this means the test site database will be repopulated between the tests.
      cleanup: true
      waitlock: 10
      url: '%TEST_SITE_WP_URL%'
      originalUrl: '%TEST_SITE_WP_URL%'
      urlReplacement: true #replace the hardcoded dump URL with the one above
      tablePrefix: '%TEST_SITE_TABLE_PREFIX%'
    \Helper\AcceptanceWPWebDriver:
      url: '%TEST_SITE_WP_URL%'
      adminUsername: '%TEST_SITE_ADMIN_USERNAME%'
      adminPassword: '%TEST_SITE_ADMIN_PASSWORD%'
      adminPath: '%TEST_SITE_WP_ADMIN_PATH%'
      port: 9515
      browser: chrome
      headers:
        X_TEST_REQUEST: 1
        X_WPBROWSER_REQUEST: 1
```

The secret sauce here is the depends clause for my new `\Helper\AcceptanceWPWebDriver` module (which we're coming to). The config for that just gets passed into the `WPWebDriver` modele which actually uses them.

The actual `Helper\AcceptanceWPWebDriver.php` file looks a bit like this:

```php
<?php

namespace Helper;

// here you can define custom actions
// all public methods declared in helper class will be available in $I

use Codeception\Exception\ModuleException;

class AcceptanceWPWebDriver extends \Codeception\Module\WPWebDriver {
	public function getLoginUrl() {
		// this is the thing I need to change about logging in for testing purposes
		return $this->loginUrl.'?wle';
	}

	/**
	 * An identical copy of the function of the same name as defined in WPWebDriver.
	 *
	 * We are overriding it since it needs to call our own getLoginUrl() from above
	 */
	public function loginAs($username, $password, $timeout = 10, $maxAttempts = 5) {

		if ($this->loginAttempt === $maxAttempts) {
			throw new ModuleException(
				__CLASS__,
				"Could not login as [{$username}, {$password}] after {$maxAttempts} attempts."
			);
		}

		$this->debug("Trying to login, attempt {$this->loginAttempt}/{$maxAttempts}...");

		$this->amOnPage($this->getLoginUrl());

		$this->waitForElement('#user_login', $timeout);
		$this->waitForElement('#user_pass', $timeout);
		$this->waitForElement('#wp-submit', $timeout);

		$this->fillField('#user_login', $username);
		$this->fillField('#user_pass', $password);
		$this->click('#wp-submit');

		$authCookie    = $this->grabWordPressAuthCookie();
		$loginCookie   = $this->grabWordPressLoginCookie();
		$empty_cookies = empty($authCookie) && empty($loginCookie);

		if ($empty_cookies) {
			$this->loginAttempt++;
			$this->wait(1);
			$this->loginAs($username, $password, $timeout, $maxAttempts);
		}

		$this->loginAttempt = 0;
	}
```

As mentioned in the comments, it is the `getLoginUrl()` method that is the important thing here. It also seems to require `loginAs()` to be defined here, otherwise the normal `loginAs()` that comes with `WpWebDriver` will just call the normal `getLoginUrl()` method instead of this one.

The other thing to note is that this class extends `\Codeception\Module\WPWebDriver` and not the standard `extends \Codeception\Module`

It took me a lot of debugging to figure this out - even though it all makes sense when you read the docs (as I had to do multiple times to figure it out). Hope this comes in handy for someone else - or even my future self!

**Comment by lucatume** [comment]
@charlesdeb thanks for sharing your findings and your solution. I will re-read this and look into what could become part of the project when I can allocate some more time to this. 

 While I tip my hat to your solution, it seems like too much job for the simple addition of the `wle` query arg to the login URL.
 
 Both the `WPBrowser` and the `WPWebDriver` modules come with an `adminPath` configuration parameter and, I think, it would make sense to add the `loginPath` parameter (default `/wp-login.php`) to allow its customization. In your case all this could be replaced by `loginPath: /wp-login.php?wle`. Correct?
 
I will post updates in this issue when I have them.

**Comment by charlesdeb** [comment]
Ha - you might just be right! I will try your suggestion and see if that works. Many thanks. It would certainly be a lot simpler than what I cooked up - although I think my original request for some more examples still stands.

BTW, I have also spent quite a bit of time trying to figure out creating extra helper methods for populating the database with fixtures with WPBrowser e.g. `$I->haveCustomPostXInDatabase($args)`. I have finally figured that out too - although I can't get my solution to also work with WPUnit tests that use the built in WP factory method that you recommend. I come from a rails background where testing is such built into everything, and while it is not always straightforward there either, I find in the PHP world that the barriers to testing are quite a bit higher. Your fine work goes a good way to lowering that barrier - but still requires a lot of time going over your docs, the codeception docs and the WP docs to get things working.
